Model/Indexer/ParallelIndexer.php

```diff
@@ -0,0 +1,209 @@
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2019 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Model\Indexer;
+
+use Magento\Framework\Indexer\DimensionalIndexerInterface;
+use Magento\Framework\Indexer\DimensionProviderInterface;
+use Magento\Framework\Indexer\Dimension;
+use Nosto\NostoException;
+use Nosto\Tagging\Model\Indexer\Data\DimensionModeConfiguration;
+use Nosto\Tagging\Helper\Account as NostoHelperAccount;
+use Nosto\Tagging\Helper\Scope as NostoHelperScope;
+use Nosto\Tagging\Logger\Logger as NostoLogger;
+use Magento\Store\Model\StoreDimensionProvider;
+use Magento\Indexer\Model\ProcessManager;
+use Nosto\Tagging\Util\Benchmark;
+use Magento\Framework\App\ObjectManager;
+use Nosto\Tagging\Model\Indexer\Dimensions\ModeSwitcherInterface;
+use Magento\Store\Model\Store;
+
+abstract class ParallelIndexer implements DimensionalIndexerInterface
+{
+    /** @var NostoHelperAccount */
+    private $nostoHelperAccount;
+
+    /** @var NostoHelperScope */
+    private $nostoHelperScope;
+
+    /** @var NostoLogger */
+    public $nostoLogger;
+
+    /** @var ProcessManager */
+    private $processManager;
+
+    /**
+     * ParallelIndexer constructor.
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoHelperScope $nostoHelperScope
+     * @param NostoLogger $nostoLogger
+     * @param ProcessManager|null $processManager
+     */
+    public function __construct(
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperScope $nostoHelperScope,
+        NostoLogger $nostoLogger,
+        ProcessManager $processManager = null
+    ) {
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->nostoHelperScope = $nostoHelperScope;
+        $this->nostoLogger = $nostoLogger;
+        $this->processManager = $processManager;
+    }
+
+    /**
+     * Get ModeSwitcher class to later get the indexer mode
+     *
+     * @return ModeSwitcherInterface
+     */
+    abstract public function getModeSwitcher(): ModeSwitcherInterface;
+
+    /**
+     * Implement logic of single store indexing
+     *
+     * @param Store $store
+     * @param array $ids
+     * @return mixed
+     */
+    abstract public function doIndex(Store $store, array $ids = []);
+
+    /**
+     * @return DimensionProviderInterface|null
+     */
+    private function getDimensionProvider()
```

Comment:
I would inject an array of "modes" and the associated "dimension provider" and remove the ObjectManager instantiation.

Reply:
The modes are based on the indexer type, so it's best for the indexer class to return the class that is responsible for it.